smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/scheduler/SmallFileScheduler.java

```diff
@@ -0,0 +1,299 @@
+import org.smartdata.hdfs.HadoopUtil;
+import org.smartdata.hdfs.action.HdfsAction;
+import org.smartdata.hdfs.action.SmallFileCompactAction;
+import org.smartdata.metastore.MetaStore;
+import org.smartdata.metastore.MetaStoreException;
+import org.smartdata.model.ActionInfo;
+import org.smartdata.model.CompactFileState;
+import org.smartdata.model.FileContainerInfo;
+import org.smartdata.model.FileInfo;
+import org.smartdata.model.FileState;
+import org.smartdata.model.LaunchAction;
+import org.smartdata.model.action.ScheduleResult;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class SmallFileScheduler extends ActionSchedulerService {
+  private final URI nnUri;
+  private DFSClient dfsClient;
+  private MetaStore metaStore;
+
+  /**
+   * container files lock, and retry number.
+   */
+  private Map<String, Integer> containerFilesLock;
+
+  /**
+   * The mapping between action and container file exist info.
+   */
+  private Map<Long, ContainerFileExistInfo> containerFileMap;
+
+  /**
+   * The mapping between action and file container information.
+   */
+  private Map<Long, Map<String, FileContainerInfo>> fileContainerInfoMap;
+
+  /**
+   * The mapping between small file and action.
+   */
+  private List<String> smallFilesLock;
+
+  /**
+   * The mapping between action and small files.
+   */
+  private Map<Long, ArrayList<String>> smallFilesMap;
+
+  private static final int MAX_RETRY_COUNT = 3;
+  private static final List<String> ACTIONS = Arrays.asList("write", "compact");
+  public static final Logger LOG = LoggerFactory.getLogger(SmallFileScheduler.class);
+
+  public SmallFileScheduler(SmartContext context, MetaStore metaStore)
+      throws IOException {
+    super(context, metaStore);
+    this.metaStore = metaStore;
+    this.nnUri = HadoopUtil.getNameNodeUri(getContext().getConf());
+  }
+
+  @Override
+  public void init() throws IOException {
+    this.containerFilesLock = new ConcurrentHashMap<>();
+    this.containerFileMap = new ConcurrentHashMap<>();
+    this.fileContainerInfoMap = new ConcurrentHashMap<>();
+    this.smallFilesLock = new ArrayList<>();
+    this.smallFilesMap = new ConcurrentHashMap<>();
+    this.dfsClient = HadoopUtil.getDFSClient(nnUri, getContext().getConf());
+  }
+
+  @Override
+  public List<String> getSupportedActions() {
+    return ACTIONS;
+  }
+
+  private class ContainerFileExistInfo {
+    private String containerFilePath;
+    private boolean isExist;
+
+    private ContainerFileExistInfo(String containerFilePath, boolean isExist) {
+      this.containerFilePath = containerFilePath;
+      this.isExist = isExist;
+    }
+  }
+
+  @Override
+  public boolean onSubmit(ActionInfo actionInfo) {
+    if (ACTIONS.get(1).equals(actionInfo.getActionName())) {
```

Comment:
Position related const value `1` may cause an error when making modifications to the array.